src/System.Private.CoreLib/shared/System/Runtime/Intrinsics/Vector128.cs

```diff
@@ -19,10 +19,158 @@ namespace System.Runtime.Intrinsics
     // This ensures we get good codegen for the "fast-path" and allows the JIT to
     // determine inline profitability of the other paths as it would normally.
 
+    // Many of the instance methods were moved to be extension methods as it results
+    // in overall better codegen. This is because instance methods require the C# compiler
+    // to generate extra locals as the `this` parameter has to be passed by reference.
+    // Having them be extension methods means that the `this` parameter can be passed by
+    // value instead, thus reducing the number of locals and helping prevent us from hitting
+    // the internal inlining limits of the JIT.
+
     public static class Vector128
     {
         internal const int Size = 16;
 
+        /// <summary>Reinterprets a <see cref="Vector128{T}" /> as a new <see cref="Vector128{U}" />.</summary>
+        /// <typeparam name="T">The type of the input vector.</typeparam>
+        /// <typeparam name="U">The type of the vector <paramref name="vector" /> should be reinterpreted as.</typeparam>
+        /// <param name="vector">The vector to reinterpret.</param>
+        /// <returns><paramref name="vector" /> reinterpreted as a new <see cref="Vector128{U}" />.</returns>
+        /// <exception cref="NotSupportedException">The type of <paramref name="vector" /> (<typeparamref name="T" />) or the type of the target (<typeparamref name="U" />) is not supported.</exception>
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<U> As<T, U>(this Vector128<T> vector)
+            where T : struct
+            where U : struct
+        {
```

Comment:
I have noticed that @terrajobst added label `api-approved`, however, could not find any link to API review meeting.

Reply:
> Anyway such wide API change should go through API review process if dotnet repos want to uphold so far strictly followed API addition/change process.

This has already undergone the necessary API review.

> Due to the fact that this method is one of the most often used 

This method is not considered to be one of the most widely used. It was only added to give users writing generic code a way to do zero-cost conversion.

Even in generic code, it is expected that the end-user will only need this when converting from a concrete type back to the generic result. When taking in generic parameters, it is expected that they will use one of the existing 10 methods to convert the generic `Vector128<T>` to a concrete type (as it is both more concise and readable).

As an example:
```csharp
Vector128<T> Add(Vector128<T> x, Vector128<T> y)
    where T : struct
{
    if (typeof(T) == typeof(float))
    {
        return Sse.Add(x.AsSingle(), y.AsSingle().As<float, T>();
    }
    else
    {
         // other types, similar code
    }
}
```

Reply:
No, it is not nearly as nice as just `As<T>()`. However, the change was made because making it (and the other helpers) instance methods was causing real-world code to hit the JIT inlining limits, as each instance call was introducing additional locals.

Reply:
Hmm, this API change is caused by some JIT optimization restrictions. Although I don’t like this kind of solutions that exposes compiler implementation details to user APIs, but the new APIs are already approved by review meeting. I don’t have strong preference …

Reply:
> causing real-world code to hit the JIT inlining limits

Was there analysis done on what it would take to fix this in the JIT instead? We prefer to not design APIs around JIT bugs.

Reply:
cc @AndyAyersMS

Reply:
@jkotas, yes. @AndyAyersMS commented here: https://github.com/dotnet/coreclr/issues/22235#issuecomment-458034471

Reply:
ok

Reply:
Just as a note: I would be perfectly happy to keep the APIs as is, if we do determine that this is fixable in the JIT (and that it could be done for the 3.0 timeframe, or shortly there-after). I don't believe there would be any pushback from the API review committee either, given this was only being done to ensure we don't ship with an easily fixable perf-bug.

It could be, for example, that the issue is more readily applicable as part of the "first class struct" work (in which case this will be something eventually addressed); it could also be that this issue something that is fairly specific to the HWIntrinsic types and it is not worth the effort in fixing.

Reply:
It's not impossible to fix this in the jit, but right now none of the options looks easy. There are a couple of avenues for exploration:
* Avoid creation of the temps in the first place, via some kind of anticipation scheme in the importer. Might be hard to make this general but could likely catch the instance-style cast operators (nullary and don't modify "this") and effectively view them as static methods.
* Tolerate creation of temps for perf sensitive code, say by upping limits if we see HW intrinsic expansions. Would need to verify this does not lead to other problems, and there are possibly multiple limits to consider.
* Reduce number of temps during or after inlining, say by growing trees or special casing temps that only live within a block (regrowing trees ~= remove temps that are only live within a block and single-def single use). Also no guarantee that temp counts will be reduced.

Reply:
Another simpler mitigation solution could be to change all the managed-implemented  `Vector128/256` helpers to be JIT intrinsic. We can just expand the helpers to their hw intrinsic implementation in importer (or codegen), which significantly reduces JIT inlining pressure and really cheap to implement. So that https://github.com/dotnet/coreclr/issues/21794 can be perfectly solved.

I am changing `Vector128/256<T>.Get/WithElement` to be intrinsics, which are not optimized by managed intrinsic yet (actually it is impossible due to https://github.com/dotnet/coreclr/issues/21794 and other factors). I will submit a PR tomorrow to demonstrate. If you think it a good way, we can change other helpers then.

Reply:
@fiigii, the issue was occurring for methods that are already treated as `JIT Intrinsics` (such as `As` and `AsSingle`). From my understanding, it has nothing to do with the method's contents or whether or not it is an intrinsic. Instead, the issue is that instance methods have to take `this` as a by-ref, which requires the C# compiler to create additional locals so it can use `ldloca` to get their address (where-as passing by value requires no-such shuffling).

The various helper methods (and only the helper methods) will also continue needing to provide software fallbacks in either case and the change that @AndyAyersMS made in https://github.com/dotnet/coreclr/pull/21893 helped greatly with the inlining issue (as now, from what I understood, only the cost of the live code-paths is considered).

In the case of `Get/WithElement`, they were not being handled by intrinsics in managed code because there isn't a good way for us to say that it is always a clear win. Instead, they likely need logic that says: "If the index is a constant, use intrinsics like `Extract/Insert/Shuffle`; otherwise, use the software fallback".

Reply:
> the issue was occurring for methods that are already treated as JIT Intrinsics (such as As and AsSingle).

Yes, but the "too many locals" are not only from "this as a by-ref", and "too many locals" is not the only problem. Currently, vector helpers heavily depend on inlining and conditional constant propagation to get the optimal codegen. But before all opt passes done, there are so many "unused" code paths that spend too much JIT budget (time, # of locals, etc.)

Reply:
> Currently, vector helpers heavily depend on inlining and conditional constant propagation to get the optimal codegen. But before all opt passes done, there are so many "unused" code paths that spend too much JIT budget (time, # of locals, etc.)

In either case, that ends up being a separate concern. This is fixing the "too many locals" issue that occurs whether the methods are treated as intrinsic or not (in many of these cases, these methods are being handled as intrinsic).

So, given the above that the alternative is not an easy fix, we likely still need to take this PR to fix the immediate issue. We can then work on treating all the helper methods as intrinsic (if needed), separately.

Reply:
For inlinees the importer is pretty good at allocating locals only "as needed." So if an inlinee has unreachable parts and some locals are referred to only in those parts, and those unreachable parts are identifiable during importation, the jit won't allocate temps for those locals (likewise for args, though that is less common).

If the unreachable parts only become evident later then yes we'll have useless temps around, and will have possibly stopped inlining "too early."

For root methods the locals are all mapped to temps up front.